bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/ReadAheadManager.java

```diff
@@ -0,0 +1,830 @@
+            }
+            readAheadTaskStatus.waitUntilReadCompleted(dbLedgerStorageStats);
+
+            entry = cache.get(ledgerId, entryId);
+            if (entry != null) {
+                if (log.isDebugEnabled()) {
+                    log.debug("[L{} E{}] Hit cache after read-ahead", ledgerId, entryId);
+                }
+                return entry;
+            }
+        }
+
+        // read the current pos entry, and add the next pos into read-ahead set
+        if (log.isDebugEnabled()) {
+            log.debug("[L{} E{}] Read from storage layer", ledgerId, entryId);
+        }
+        return readAndAddNextReadAheadPosition(ledgerId, entryId);
+    }
+
+    private void internalReadAhead(long ledgerId, long entryId, long location,
+                                   int maxEntries, int maxBytes, long submitStartNanos) {
+        // record queue time
+        recordStatsInNano(dbLedgerStorageStats.getReadAheadAsyncQueueTime(), submitStartNanos);
+        long readAheadStartNanos = MathUtils.nowInNano();
+
+        // obtain the read-ahead info
+        ReadAheadTaskStatus readAheadTaskStatus = getNearestTask(ledgerId, entryId);
+        if (readAheadTaskStatus == null) {
+            log.info("The RA-task L{} E{} has already been cleaned", ledgerId, entryId);
+            return;
+        }
+
+        // record all the positions of these entries
+        List<Long> entriesPos = new ArrayList<>();
+
+        // read from fc
+        int entries = 0;
+        int bytes = 0;
+
+        // flag to determine whether keeps reading ahead
+        boolean earlyExit = false;
+
+        // flag to indicate whether reads from cache since last time
+        boolean readFromCacheBefore = false;
+
+        try {
+            while (entries <= maxEntries && bytes <= maxBytes) {
+
+                ByteBuf entry = cache.get(ledgerId, entryId);
+                if (entry != null) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("[L{} E{}] Entry already exists.", ledgerId, entryId);
+                    }
+                    readFromCacheBefore = true;
+                    entryId++;
+                    location += (4 + entry.readableBytes());
+                    entry.release();
+                    continue;
+                }
+
+                if (readFromCacheBefore) {
+                    try {
+                        entry = entryLogger.readEntry(ledgerId, entryId, location);
+                        readFromCacheBefore = false;
+                    } catch (Exception e) {
+                        if (log.isDebugEnabled()) {
+                            log.debug("[L{} E{}] Failed to locate entry with location:{}", ledgerId, entryId, location);
+                        }
+                        earlyExit = true;
+                        break;
+                    }
+                } else {
+                    entry = entryLogger.readEntry(ledgerId, entryId, location);
```

Comment:
1. In accordance with the previous logic, we'd better use 
`entry = entryLogger.internalReadEntry(ledgerId, entryId, location, false);` don't do entry validation here.